### PR TITLE
Two-Stage SRF: velocity-first then pressure correction

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Split the SRF into two sequential stages: Stage 1 predicts (ΔUx, ΔUy) velocity corrections, Stage 2 receives backbone features + Stage 1 velocity corrections to predict Δp. This exploits Bernoulli physics: pressure is a function of velocity, so having an intermediate velocity estimate as context should improve pressure predictions.

Motivated by the success of \`--pressure_first --pressure_deep\` which already showed that separating pressure from velocity prediction helps. This takes it further: instead of predicting pressure first (which worked), we predict velocity first and use it as context for pressure. The SIMPLE algorithm in CFD solves pressure from velocity — this mirrors that physics.

### Changes to `cfd_tandemfoil/train.py`

1. **Add flag:**
   ```python
   parser.add_argument('--srf_two_stage', action='store_true')
   ```

2. **Implement TwoStageSRF:**
   ```python
   class TwoStageSRF(nn.Module):
       def __init__(self, d_input, d_hidden=192, n_layers=3):
           super().__init__()
           # Stage 1: backbone features → velocity corrections (Ux, Uy)
           layers_vel = [nn.Linear(d_input, d_hidden), nn.GELU()]
           for _ in range(n_layers - 1):
               layers_vel.extend([nn.Linear(d_hidden, d_hidden), nn.GELU()])
           layers_vel.append(nn.Linear(d_hidden, 2))  # output: ΔUx, ΔUy
           self.vel_head = nn.Sequential(*layers_vel)
           
           # Stage 2: backbone features + vel corrections → pressure
           layers_p = [nn.Linear(d_input + 2, d_hidden), nn.GELU()]
           for _ in range(n_layers - 1):
               layers_p.extend([nn.Linear(d_hidden, d_hidden), nn.GELU()])
           layers_p.append(nn.Linear(d_hidden, 1))  # output: Δp
           self.p_head = nn.Sequential(*layers_p)
       
       def forward(self, x):
           vel = self.vel_head(x)           # [B, N, 2]
           p = self.p_head(torch.cat([x, vel.detach()], -1))  # [B, N, 1]
           # detach vel so Stage 2 doesn't backprop through Stage 1
           return torch.cat([vel, p], dim=-1)  # [B, N, 3]
   ```

3. **Replace SurfaceRefinementHead** with TwoStageSRF when flag is set. Same for AftFoilSRF.

4. **Important:** Use `.detach()` on vel input to Stage 2 — this prevents Stage 2 gradients from disrupting Stage 1's velocity learning. Stage 1 trains purely on velocity loss; Stage 2 trains on pressure loss with velocity context.

5. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent frieren --wandb_name "frieren/two-stage-srf-s42" --wandb_group two-stage-srf \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
     --srf_two_stage

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent frieren --wandb_name "frieren/two-stage-srf-s73" --wandb_group two-stage-srf \
     --seed 73 [same flags]
   ```

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Experiment: Two-Stage SRF — velocity-first then pressure correction (2 seeds)

**Architecture:** TwoStageSRF with hidden=128, layers=2 (using config defaults). Stage 1: backbone features + base_pred → ΔUx, ΔUy. Stage 2: backbone features + base_pred + detached velocity corrections → Δp. Zero-initialized output layers for both stages. Total: 84,867 SRF params (vs 51,459 for standard SRF).

**Baseline** (PR #2319, 2-seed avg):
| Metric | Baseline |
|--------|----------|
| p_in | 11.709 |
| p_oodc | 7.544 |
| p_tan | 27.402 |
| p_re | 6.481 |

**Seed 42** — W&B run: `iizetcor`, best epoch 159, wall time 180.9 min
**Seed 73** — W&B run: `f4xi8sym`, best epoch 160, wall time 180.7 min

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| val/loss (in-dist) | 0.3652 | 0.3761 | 0.3707 | — | — |
| **p_in (mae_surf_p)** | **11.8** | **12.2** | **12.0** | 11.709 | **+2.5%** |
| val/loss (ood_cond) | 0.4306 | 0.4263 | 0.4285 | — | — |
| **p_oodc (mae_surf_p)** | **7.5** | **7.3** | **7.4** | 7.544 | **-1.9%** ✅ |
| val/loss (tandem) | 1.1888 | 1.1758 | 1.1823 | — | — |
| **p_tan (mae_surf_p)** | **28.6** | **27.3** | **27.95** | 27.402 | **+2.0%** |
| val/loss (ood_re) | 0.4185 | 0.4139 | 0.4162 | — | — |
| **p_re (mae_surf_p)** | **6.5** | **6.4** | **6.45** | 6.481 | **-0.5%** ✅ |

Peak VRAM: ~40.4 GB (seed 42), ~40.5 GB (seed 73)
Training speed: ~67 s/epoch (~159–160 epochs in 180 min — identical to baseline epoch count)

**Commands used:**
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/two-stage-srf-s42" --wandb_group two-stage-srf \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_two_stage

cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent frieren --wandb_name "frieren/two-stage-srf-s73" --wandb_group two-stage-srf \
  --seed 73 [same flags]
```

---

### What happened

**Mixed result — below merge threshold.** The two-stage SRF shows improvements on p_oodc (-1.9%) and p_re (-0.5%), but regressions on p_in (+2.5%) and p_tan (+2.0%). The aggregate performance does not beat the baseline.

**Analysis:**

1. **The velocity→pressure cascade works mechanically but is information-limited.** The Stage 2 pressure head receives the 2-element velocity correction concatenated with a ~128-dim backbone feature vector + 3-dim base_pred. The velocity correction is a tiny fraction of the total input (2 out of ~133 dims), so its influence on the pressure prediction is minimal. This matches the researcher-agent's concern that "the 2-element velocity correction may be too small to influence a 192-dim hidden layer."

2. **The `.detach()` prevents co-adaptation.** Stage 1 can only optimize for velocity accuracy. It cannot learn to produce velocity estimates that are particularly informative for pressure prediction. In the SIMPLE algorithm, the pressure-velocity coupling is iterative — here we get one shot with detached gradients, so Stage 2 is essentially just getting an extra 2D feature that happens to be a velocity correction.

3. **More parameters but not better.** TwoStageSRF has 84,867 params (65% more than standard SRF's ~51K). The extra capacity doesn't translate to better performance because the two heads are structurally constrained: Stage 1 can only learn 2-dim output (velocity), Stage 2 can only learn 1-dim output (pressure), vs the standard SRF which has full flexibility over all 3 dims.

4. **No training speed penalty.** At ~67 s/epoch, TwoStageSRF runs at the same speed as the baseline (both getting ~160 epochs in 180 min), confirming the two-stage computation adds negligible overhead.

5. **p_oodc improvement is promising.** The -1.9% improvement on OOD conditions suggests that the velocity→pressure inductive bias helps on extrapolation tasks, even if the detached gradients limit its effectiveness on in-distribution.

---

### Suggested follow-ups

1. **Remove `.detach()`** — allow end-to-end gradients through both stages. This lets Stage 1 learn "pressure-friendly" velocity patterns. The risk is gradient interference, but it's worth testing. This is the single highest-impact change.
2. **Project velocity corrections to higher dim** — before concatenating to Stage 2 input, pass the 2D correction through `nn.Linear(2, hidden_dim//4)` to give it more signal weight in the combined feature vector.
3. **Use hidden=192, layers=3** — the PR's class definition suggested these larger dims. The current run used config defaults (128/2) which may have been undersized for a two-stage architecture.
4. **Iterative refinement** — run Stage 1→Stage 2 multiple times (2-3 iterations), matching the iterative nature of SIMPLE more closely.
5. **Standard SRF + pressure-only second stage** — instead of splitting vel/pressure, use the existing SRF for all 3 channels, then add a pressure-only refinement head on top. This preserves the flexibility of the standard SRF while adding pressure-specific correction.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2355 because ADVISOR: Closing — mixed results (p_in +2.5%, p_oodc -1.9%, p_tan +2.0%, p_re -0.5%) net negative against the current baseline. The two-stage approach (velocity-first then pressure) adds training complexity without consistent improvement. The current joint training with PCGrad already handles the multi-task balance well, and splitting stages reduces the effective training time for each output. Thank you for the clean implementation, frieren..
